[REVIEW] Integrate dask-cugraph in cugraph

afender opened the pull request:
Close #396 
- Upgraded cugrpah conda env and CI script
- Changed dask-cugraph organization to match cugraph
- Updated core.py with latest cuml changes
- Make all the code flake8 compliant

Close #404 
- added python doc

Extras :
- Fixed karate test
- Deleted irrelevant pytest warning (coming from dependencies)

seunghwak commented on docs/source/api.rst:
```diff
@@ -62,3 +62,13 @@ Sssp
+
+.. automodule:: cugraph.dask.pagerank.pagerank
+    :members:
+    :undoc-members:
```
This will conflict with the documentation updates (https://github.com/rapidsai/cugraph/pull/412). Had a chance to build a new document? We need to decide how to present SNMG algorithms in the documentation. If you want to merge this faster, we may leave this SNMG routine as undocumented, and resolve this in a separate issue (and discuss what's the proper structure to document SNMG routines there).

And side notes, the number of "==" and "-" under Dask and Pagerank should match the number of characters in Dask and Pagerank (if not, you'll see warnings when you run make html).

afender replied:
Yeah, I was trying to follow the syntax on #412, I guess I didn't get all the subtilities yet :). 
I was hoping to merge this today to simplify the development of dask related features to everyone involved. 

Do I understand correctly that you recommend dropping my change in `docs/source/api.rst`, merge both #412 and #411 asap without conflicts, and address the dask update in `docs/source/api.rst` later?

seunghwak replied:
Yes, to merge this faster, we may just drop changes in api.rst in this PR. #412 should be merged before we start implementing SNMG documentations.

afender replied:
dropped doc in api.rst

seunghwak commented on python/cugraph/dask/core.py:
```diff
@@ -0,0 +1,184 @@
+import time
+import os
+from threading import Lock, Thread
+import cugraph
```
Do we have any guidelines on import orders? In this file cugraph is imported last and in the pagerank.py below, it's imported first. Not a critical thing for this PR, but enforcing some consistency will help code readability.

afender replied:
This order is the proper one. 

According to PEP8 :
```
Imports should be grouped in the following order:
Standard library imports.
Related third party imports.
Local application/library specific imports.
````
https://www.python.org/dev/peps/pep-0008/#imports

Notice that it is pretty close to what the order of C++ header should be ideally.

seunghwak replied:
Great, and then, it might be great if we can at least encourage people to follow this order (enforcing may be too much for some people).

afender replied:
Agreed :) Let's bring that to the next sync meeting. This would fit into the refactoring document too.

afender replied:
google doc has been updated

seunghwak commented on python/cugraph/dask/core.py:
```diff
@@ -0,0 +1,184 @@
+            logging.error("Error opening ipc_handle on device " +
+                          str(self.device) + ": " + str(e))
+
+        self.lock.release()
```
What happens to the lock if exception occurs? If the lock stays in the acquired state, doesn't this create any problem? I don't understand the entire context, but if this is problematic, self.lock.release() should go to the finally part (in try-except-finally).

afender replied:
We borrowed `core.py` from cuML. I think @cjnolet is the original author? Or maybe @Iroy30 knows?

seunghwak commented on python/cugraph/dask/core.py:
```diff
@@ -0,0 +1,184 @@
+        self.lock.release()
+
+        while (self.running):
+            time.sleep(0.0001)
```
May adding a comment explaining why you picked 0.0001 will be helpful.

afender replied:
I wish this file was in raft so that we address that once. @Iroy30 any thought?

Iroy30 replied:
To have a while loop continuously checking on self.running status. time 0.0001 is probably an arbitrary small enough time to repeat the condition checking. In cpp, probably a "while(condition);" is enough but python would need to have a function body

seunghwak commented on python/cugraph/dask/core.py:
```diff
@@ -0,0 +1,184 @@
+        place underneath.
+        """
+        while (not self.running):
+            time.sleep(0.0001)
```
Maybe adding a comment explaining why you picked 0.0001 will be helpful.

seunghwak commented on python/cugraph/dask/core.py:
```diff
@@ -0,0 +1,184 @@
+
+def select_device(dev, close=True):
+    """
+    Use numbas numba to select the given device, optionally
```
Use numbas numba to select => Use numba to select?

afender replied:
another one in `core.py` I wonder if we should just do address that as part of refactoring and raft transition

afender replied:
(so that we core.py remains consistent in cugraph/cuml in the meantime)

seunghwak commented on python/cugraph/dask/pagerank/pagerank.py:
```diff
@@ -0,0 +1,253 @@
+    Parameters
+    ----------
+    edge_list : dask_cudf.DataFrame
+       Contain the connectivity information as an edge list
```
Missing period at the end. And it seems like using 4 spaces is most common in cugraph, cudf, & cuml docstrings (here, 3 spaces are used).

afender replied:
fixed

seunghwak commented on python/cugraph/dask/pagerank/pagerank.py:
```diff
@@ -0,0 +1,253 @@
+    --------
+    >>> import dask_cugraph.pagerank as dcg
+    >>> chunksize = dcg.get_chunksize(edge_list.csv)
+    >>> ddf_edge_list = dask_cudf.read_csv(edge_list.csv,
```
Does this example run? No need to import dask_cudf?

afender replied:
No, it is just a snapshot of the relevant part (like in most cugraph doc). I think that is good enough to understand the API. We will have a notebook for a working example.

seunghwak replied:
I personally believe giving a running example will be helpful for the audiences who want to try examples (I will be upset if I read a documentation, type what appears there, and see error messages), but I will let you make the verdict.

seunghwak commented on python/cugraph/dask/pagerank/pagerank.py:
```diff
@@ -0,0 +1,253 @@
+    return gpu_data
+
+
+def get_n_gpus():
```
get_num_gpus or get_number_of_gpus might be clearer. I wasn't sure what this function is just seeing the function name.

Iroy30 replied:
that is consistent with dask-cudas get_n_gpus function

afender replied:
Resolved then. We should be consist with existing dask-* packages

seunghwak commented on python/cugraph/tests/dask/__init__.py:
```diff
@@ -0,0 +1,15 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
```
What is this? Maybe something coming from a specific editor?

afender replied:
fixed

seunghwak commented on python/cugraph/tests/dask/test_hibench_small.py:
```diff
@@ -0,0 +1,57 @@
+        if(abs(res_df['pagerank'][i]-nx_pr[i][1]) > tol*1.1):
+            err = err + 1
+    print("Mismatches:", err)
+    assert err < (0.02*len(res_df))
```
client.close()
cluster.close()

If not, if run this as a stand-alone python program, you will see some warnings saying processes are killed.

afender replied:
good catch. fixed

seunghwak commented on python/cugraph/tests/dask/test_karate_csv.py:
```diff
@@ -0,0 +1,49 @@
+        if(abs(res_df['pagerank'][i]-nx_pr[i][1]) > tol*1.1):
+            err = err + 1
+    print("Mismatches:", err)
+    assert err < (0.01*len(res_df))
```
client.close()
cluster.close()

here as well.

afender replied:
good catch. fixed

afender commented:
> Looks good to me besides what I have commented.

@seunghwak I addressed all your comments except those in `core.py`. 

We may want to keep `core.py` consistent with cuML for now. Since cuML is the original owner, I would like to be able to easily replace it with the new version when they add bug fixes. While your comments are valid, they don't point to flaws that would prevent running the main feature. I'd say we keep `core.py` it as is and keep working offline with cuML to best address core.py maintenance. Any objection?

seunghwak commented:
> > Looks good to me besides what I have commented.
> 
> @seunghwak I addressed all your comments except those in `core.py`.
> 
> We may want to keep `core.py` consistent with cuML for now. Since cuML is the original owner, I would like to be able to easily replace it with the new version when they add bug fixes. While your comments are valid, they don't point to flaws that would prevent running the main feature. I'd say we keep `core.py` as is and keep working offline with cuML to best address core.py maintenance. Any objection?

I didn't know it's from cuML, and sounds reasonable if it will eventually go to raft.

afender commented:
rerun tests
